### PR TITLE
Update concatdataset

### LINKED ISSUE #32034: Extend `ConcatDataset` to return dataset index
## 🚀 Feature
This feature would extend  `torch.utils.data.ConcatDataset` to return the data set index from which a sample originated.

## Motivation

This would be useful when the user wants to keep track of which data set each sample comes from.

The use case I have in mind is multi-task learning when using data sets that contain mutually exclusive labels (e.g. dataset A contains labels for Task A, and dataset B contains labels for task B) as opposed to all labels being present in a single data set. 

In this case, it is still possible to do multi-task learning by pooling examples from different source, if one keeps track of which source each example in a batch came from, which is the motivation behind this request.

## Pitch
The feature would consist of two changes:

* adding an extra `return_index` argument to the class (defaulting to `False`)
* adding `dataset_idx` to the return if `return_index` is set to `True` (https://github.com/pytorch/pytorch/blob/8ea49e7a0897b77a2cd6bf854efd80ef593acf02/torch/utils/data/dataset.py#L207)

### PR DESCRIPTION
**Pull request by ATriantafyllopoulos**
This PR restructures `ConcatDataset` as per the discussion in #32034, specifically according to [this comment](https://github.com/pytorch/pytorch/issues/32034#issuecomment-573016452)

It introduces a new `get_idxs` that returns the `dataset` and `sample` indices, which are then used by `__getitem__` to return the actual sample.

Resolves #32034

### CODE REVIEW

File: torch/utils/data/dataset.py [hunk 1]
```diff
@@ -194,16 +194,21 @@ def __init__(self, datasets):
     def __len__(self):
         return self.cumulative_sizes[-1]
 
-    def __getitem__(self, idx):
-        if idx < 0:
-            if -idx > len(self):
-                raise ValueError("absolute value of index should not exceed dataset length")
-            idx = len(self) + idx
+    def get_idxs(self, idx):
```
Comment (zou3519) [inline torch/utils/data/dataset.py:197]: Can you add a test to https://github.com/pytorch/pytorch/blob/13120bf67742e01d5e047b04f3b1e29e0a313b80/test/test_dataloader.py#L254 ?

Comment (zou3519) [inline torch/utils/data/dataset.py:197]: We should add a docstring to this so that users know what it does.

Comment (SsnL) [reply torch/utils/data/dataset.py:197]: this needs a much better name too

Comment (ATriantafyllopoulos) [reply torch/utils/data/dataset.py:197]: Unfortunately I am not too familiar with naming conventions in `PyTorch`, so I tried to keep it short.

I prefer more verbose names, e.g. `get_dataset_sample_idxs` or something similar..Anything better comes to mind?

P.S. I will add the test too ofc

Comment (ATriantafyllopoulos) [reply torch/utils/data/dataset.py:197]: e8c095f  introduced a test for `get_idxs`

I am not so sure about f4d924e

`get_idxs` will not give an `IndexError` if the user requires an index that's out of bounds, so added a failure case for when the user actually tries to use the returned index on the data set itself. Any opinions of how to actually handle this?

Comment (ATriantafyllopoulos) [reply torch/utils/data/dataset.py:197]: Added a docstring in 718999e

**Comment by tejank10** [comment]
Bump! Any plans for merging this? This feature be quite useful! 
cc @ATriantafyllopoulos @apaszke

**Comment by ATriantafyllopoulos** [comment]
> Bump! Any plans for merging this? This feature could be quite useful, thanks!
> cc @ATriantafyllopoulos @apaszke

I would be happy to rebase on the master and continue. I would first need some feedback on the open questions..
